test/sagemaker_tests/__init__.py

```diff
@@ -115,43 +145,105 @@ def invoke_sm_helper_function(ecr_image, sagemaker_regions, test_function, *test
             return
         except sagemaker.exceptions.UnexpectedStatusException as e:
             if "CapacityError" in str(e):
+                error = e
                 continue
             else:
                 raise e
-
-
+        except botocore.exceptions.ClientError as e:
+            if "ThrottlingException" in str(e) or "ResourceLimitExceeded" in str(e):
+                error = e
+                continue
+            else:
+                raise e
+    if "CapacityError" in str(error):
+        raise SMInstanceCapacityError from error
+    elif "ResourceLimitExceeded" in str(error):
+        raise SMResourceLimitExceededError from error
+    else:
+        raise SMThrottlingError from error
+
+
+@retry(
+    reraise=True,
+    retry=retry_if_exception_type(
+        (SMInstanceCapacityError, SMResourceLimitExceededError, SMThrottlingError)
+    ),
+    stop=stop_after_delay(20 * 60),
+    wait=wait_fixed(60),
+)
 def invoke_sm_endpoint_helper_function(
     ecr_image,
     sagemaker_regions,
-    local_model_path,
-    model_helper,
     test_function,
+    local_model_paths=None,
+    model_helper=None,
+    mme_folder_name=None,
+    dump_logs_from_cloudwatch=None,
     **test_function_args,
 ):
     ecr_image_region = get_ecr_image_region(ecr_image)
+    error = None
     for region in sagemaker_regions:
         sagemaker_client = get_sagemaker_client(region)
         boto_session = boto3.Session(region_name=region)
         sagemaker_runtime_client = get_sagemaker_runtime_client(region)
+        sagemaker_session = Session(boto_session=boto3.Session(region_name=region))
         # Reupload the image to test region if needed
         tested_ecr_image = (
             get_ecr_image(ecr_image, region) if region != ecr_image_region else ecr_image
         )
-        tested_model_data = model_helper(
-            region=region, boto_session=boto_session, local_path=local_model_path
-        )
+
+        # Run model setup if needed
+        tested_model_data = None
+        if model_helper and local_model_paths:
+            if not mme_folder_name:
+                tested_model_data = model_helper(
+                    region=region, boto_session=boto_session, local_path=local_model_paths[0]
+                )
+            else:
+                tested_model_data = model_helper(
+                    region=region,
+                    boto_session=boto_session,
+                    mme_folder_name=mme_folder_name,
+                    path_list=local_model_paths,
+                )
```

Comment:
While reading the entire PR, I realized that the reason could be that this method needs to be much more generic now since it is being used by various tests.

Reply:
Yeah, the behavior of `model_helper` is different for MME tests and regular TFS endpoint tests.

And in non-TF inference tests, this whole model_helper section is completely unnecessary. Since I'm reusing this function to run SM inference tests for all frameworks, I've tried to make it capable of handling all scenarios.